File: Mandatory/004 - Palindrome Index/PalindromeIndex.py
def isPalindromeArray(arr):
  reverseArr = list(map(lambda item: item, arr))
  reverseArr.reverse()
  isPalindrome = (''.join(list(map(lambda item: str(item), arr))) == ''.join(list(map(lambda item: str(item), reverseArr))))
  return isPalindrome

def palindromeIndex(s):
  charsArray = list(s)
  stringLength = len(charsArray)
  index = 0
  banExit = False
  
  while ((not banExit) and (index < stringLength - index - 1)):
    if (charsArray[index] == charsArray[stringLength - index - 1]):
      index += 1
    else:
      banExit = True
      temp1Array = list(map(lambda item: item, charsArray))
      temp1Array.pop(index)
      if (not isPalindromeArray(temp1Array)):
        temp2Array = list(map(lambda item: item, charsArray))
        temp2Array.pop(stringLength - index - 1)  
        if (isPalindromeArray(temp2Array)):          
          index = stringLength - index - 1
        else:
          index = -1
            
  if (banExit): 
    return index
  
  return -1

print(palindromeIndex("sergio"))
print(palindromeIndex("axxa"))
print(palindromeIndex("aaab"))
print(palindromeIndex("baa"))

File: Mandatory/004 - Palindrome Index/PalindromeIndex.rb
def isPalindromeArray(arr)
  reverseArr = arr.map { |item| item }.reverse()
  isPalindrome = (arr.join('') == reverseArr.join(''))
  isPalindrome
end

def palindromeIndex(s)
  charsArray = s.split('')
  stringLength = charsArray.length
  index = 0
  banExit = false

  while !banExit && index < stringLength - index - 1 do
    if (charsArray[index] == charsArray[stringLength - index - 1])
      index += 1
    else
      banExit = true
      temp1Array = charsArray.map { |item| item }
      temp1Array.delete_at(index)
      if !isPalindromeArray(temp1Array)
        temp2Array = charsArray.map { |item| item }
        temp2Array.delete_at(stringLength - index - 1)
        if isPalindromeArray(temp2Array)
          index = stringLength - index - 1
        else
          index = -1
        end
      end
    end
  end

  return index if banExit
  -1
end
   
p palindromeIndex('sergio')
p palindromeIndex('axxa')
p palindromeIndex('aaab')
p palindromeIndex('baa')
